src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs

```diff
@@ -335,29 +331,34 @@ string IHttpConnectionFeature.ConnectionId
             set
             {
                 _clientCert = value;
+                _clientCertTask = Task.FromResult(value);
                 SetInitialized(Fields.ClientCertificate);
             }
         }
 
-        async Task<X509Certificate2?> ITlsConnectionFeature.GetClientCertificateAsync(CancellationToken cancellationToken)
+        Task<X509Certificate2?> ITlsConnectionFeature.GetClientCertificateAsync(CancellationToken cancellationToken)
         {
-            if (IsNotInitialized(Fields.ClientCertificate))
+            if (_clientCertTask != null)
             {
-                var method = Server.Options.ClientCertificateMethod;
-                if (method != ClientCertificateMethod.NoCertificate)
-                {
-                    // Check if a cert was already available on the connection.
-                    _clientCert = Request.ClientCertificate;
-                }
+                return _clientCertTask;
+            }
 
-                if (_clientCert == null && method == ClientCertificateMethod.AllowRenegotation)
-                {
-                    _clientCert = await Request.GetClientCertificateAsync(cancellationToken);
-                }
+            var tlsFeature = (ITlsConnectionFeature)this;
+            var clientCert = tlsFeature.ClientCertificate; // Lazy initialized
+            if (clientCert != null
+                || Server.Options.ClientCertificateMethod != ClientCertificateMethod.AllowRenegotation
```

Comment:
ClientCertificateMethod.AllowRenegotation == ClientCertificateMode.DelayCertificate?

Reply:
Yes, Kestrel and HttpSys use separate enums. We didn't re-use the old name after learning the mechanics have changed in TLS 1.3 to get rid of renegotiation and add an alternate feature just for client certs.